lib/Search/AOD/LuceneSearchEngine.php

```diff
@@ -0,0 +1,178 @@
+    private $index;
+
+    public function __construct()
+    {
+        $this->index = BeanFactory::getBean("AOD_Index")->getIndex();
+    }
+
+    /**
+     * Search function run when user goes to Show All and runs a search again.  This outputs the search results
+     * calling upon the various listview display functions for each module searched on.
+     *
+     * @param SearchQuery $query
+     *
+     * @return SearchResults
+     * @throws Exception
+     */
+    public function search(SearchQuery $query): SearchResults
+    {
+        $queryString = $query->getSearchString();
+
+        $start = microtime(true);
+        $hits = $this->runLucene($queryString);
+        $results = $this->parseHits($hits);
+        $end = microtime(true);
+        $elapsed = $end - $start;
+
+        return new SearchResults($results['modules'], true, $elapsed, count($results['hits']));
+    }
+
+    /**
+     *
+     * @param string $queryString
+     * @return array
+     */
+    private function runLucene(string $queryString): array
+    {
+        $cachePath = 'cache/modules/AOD_Index/QueryCache/' . md5($queryString);
+        if (is_file($cachePath)) {
+            $mTime = filemtime($cachePath);
+            if ($mTime > (time() - 5 * 60)) {
+                $hits = unserialize(sugar_file_get_contents($cachePath));
+            }
+        }
+
+        if (!isset($hits)) {
+            $hits = $this->newHit($queryString);
+        }
+
+        return $hits;
+    }
+
+    /**
+     * @param string $queryString
+     * @return array
+     */
+    private function newHit(string $queryString): array
+    {
+        global $current_user;
+
+        $tmphits = $this->index->find($queryString);
+        $hits = [];
+        foreach ($tmphits as $hit) {
+            $bean = BeanFactory::getBean($hit->record_module, $hit->record_id);
+            if (empty($bean)) {
+                continue;
+            }
+            if ($bean->bean_implements('ACL') && !is_admin($current_user)) {
+                $in_group = SecurityGroup::groupHasAccess($bean->module_dir, $bean->id, 'list');
+                $is_owner = $bean->isOwner($current_user->id);
+                $access = ACLController::checkAccess($bean->module_dir, 'list', $is_owner, 'module', $in_group);
+                if (!$access) {
+                    continue;
+                }
+            }
+            $newHit = new stdClass;
+            $newHit->record_module = $hit->record_module;
+            $newHit->record_id = $hit->record_id;
+            $hits[] = $newHit;
+        }
+        $this->cacheQuery($queryString, $hits);
+
+        return $hits;
+    }
+
+    /**
+     * @param mixed $hits
+     * @return array
+     */
+    private function parseHits(array $hits): array
+    {
+        $searchResults = [];
+
+        foreach ($hits as $i => $array) {
+            $recordModule = $array->record_module;
+            $searchResults[$recordModule][$i] = $array->record_id;
```

Comment:
Hi @Dillon-Brown, does the index in `$searchResults[$recordModule][$i]` matter? because using `$i` for the index may lead to gaps/jumps in the indexes. If we want incremental indexes without jumps we could just do`$searchResults[$recordModule][]  = $array->record_id;`

Reply:
Nope it doesn't matter, good point. Pushed.

Reply:
Nope it doesn't matter, good point. Pushed.

Reply:
thank you